File: 2026美赛高频算法+创新算法汇总+源代码_副本/50多种常用算法源代码！（Matlab+Python）/标准差法检测异常值Matlab+Python代码/std_outlier_detection.m
% 标准差法检测异常值案例说明：
% 本案例使用标准差法检测数据中的异常值。标准差法假设数据服从正态分布，
% 通常将偏离均值超过3个标准差的数据点视为异常值（3σ原则）。
% 示例中生成服从正态分布的模拟数据并添加异常值，使用3σ原则进行检测。

% 设置随机种子，确保结果可复现
rng(42);

% 生成示例数据：正态分布数据 + 异常值
mu = 50;                  % 均值
sigma = 10;               % 标准差
normal_data = normrnd(mu, sigma, 300, 1);  % 正常数据
outliers = [10; 20; 90; 100; 110; 120];    % 异常值
data = [normal_data; outliers];            % 合并数据

% 计算均值和标准差
mean_val = mean(data);
std_val = std(data);

% 定义异常值阈值（3σ原则）
lower_threshold = mean_val - 3 * std_val;
upper_threshold = mean_val + 3 * std_val;

% 检测异常值
outlier_indices = find((data < lower_threshold) | (data > upper_threshold));
outlier_values = data(outlier_indices);
normal_values = data(find((data >= lower_threshold) & (data <= upper_threshold)));

% 输出结果
fprintf('标准差法(3σ原则)异常值检测结果：\n');
fprintf('数据总量: %d 个\n', length(data));
fprintf('均值(μ): %.2f\n', mean_val);
fprintf('标准差(σ): %.2f\n', std_val);
fprintf('异常值下限: μ - 3σ = %.2f\n', lower_threshold);
fprintf('异常值上限: μ + 3σ = %.2f\n', upper_threshold);
fprintf('检测到异常值数量: %d 个\n', length(outlier_values));
fprintf('异常值: ');
for i = 1:length(outlier_values)
    fprintf('%.2f ', outlier_values(i));
end
fprintf('\n');

% 可视化结果
figure;
edges = linspace(min(data), max(data), 31);  % 直方图区间
[N, bins] = histcounts(data, edges);

% 绘制直方图并区分正常数据和异常值
for i = 1:length(N)
    bin_center = (bins(i) + bins(i+1)) / 2;
    if bin_center < lower_threshold || bin_center > upper_threshold
        bar(bin_center, N(i), (bins(i+1)-bins(i))*0.9, 'FaceColor', 'r', 'EdgeColor', 'k');
    else
        bar(bin_center, N(i), (bins(i+1)-bins(i))*0.9, 'FaceColor', 'b', 'EdgeColor', 'k');
    end
    hold on;
end

% 绘制均值和3σ线
line([mean_val mean_val], [0 max(N)], 'Color', 'g', 'LineWidth', 2);
line([lower_threshold lower_threshold], [0 max(N)], 'Color', 'orange', 'LineStyle', '--', 'LineWidth', 2);
line([upper_threshold upper_threshold], [0 max(N)], 'Color', 'orange', 'LineStyle', '--', 'LineWidth', 2);

% 添加标签和图例
title('标准差法(3σ原则)异常值检测');
xlabel('数据值');
ylabel('频数');
legend('正常数据', '异常值', sprintf('均值 (μ = %.2f)', mean_val), ...
       sprintf('3σ 边界'), 'Location', 'best');
grid on;
hold off;
    

File: 2026美赛高频算法+创新算法汇总+源代码_副本/50多种常用算法源代码！（Matlab+Python）/标准差法检测异常值Matlab+Python代码/std_outlier_detection.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
标准差法检测异常值案例说明：
本案例使用标准差法检测数据中的异常值。标准差法假设数据服从正态分布，
通常将偏离均值超过3个标准差的数据点视为异常值（3σ原则）。
示例中生成服从正态分布的模拟数据并添加异常值，使用3σ原则进行检测。
"""

import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches

# 生成示例数据：正态分布数据 + 异常值
np.random.seed(42)  # 设置随机种子，确保结果可复现
mu = 50             # 均值
sigma = 10          # 标准差
normal_data = np.random.normal(mu, sigma, 300)  # 正常数据
outliers = np.array([10, 20, 90, 100, 110, 120])  # 异常值
data = np.concatenate([normal_data, outliers])    # 合并数据

# 计算均值和标准差
mean = np.mean(data)
std = np.std(data)

# 定义异常值阈值（3σ原则）
lower_threshold = mean - 3 * std
upper_threshold = mean + 3 * std

# 检测异常值
outlier_indices = np.where((data < lower_threshold) | (data > upper_threshold))[0]
outlier_values = data[outlier_indices]
normal_values = data[np.where((data >= lower_threshold) & (data <= upper_threshold))[0]]

# 输出结果
print("标准差法(3σ原则)异常值检测结果：")
print(f"数据总量: {len(data)} 个")
print(f"均值(μ): {mean:.2f}")
print(f"标准差(σ): {std:.2f}")
print(f"异常值下限: μ - 3σ = {lower_threshold:.2f}")
print(f"异常值上限: μ + 3σ = {upper_threshold:.2f}")
print(f"检测到异常值数量: {len(outlier_values)} 个")
print(f"异常值: {', '.join([f'{x:.2f}' for x in outlier_values])}")

# 可视化结果
plt.figure(figsize=(12, 6))

# 绘制直方图
n, bins, patches = plt.hist(data, bins=30, alpha=0.7, color='skyblue', edgecolor='black')

# 标记正常数据和异常值区域
for patch in patches:
    x = patch.get_x() + patch.get_width()/2
    if x < lower_threshold or x > upper_threshold:
        patch.set_facecolor('red')
    else:
        patch.set_facecolor('skyblue')

# 绘制均值和3σ线
plt.axvline(mean, color='green', linestyle='-', linewidth=2, label=f'均值 (μ = {mean:.2f})')
plt.axvline(lower_threshold, color='orange', linestyle='--', linewidth=2, 
           label=f'μ - 3σ = {lower_threshold:.2f}')
plt.axvline(upper_threshold, color='orange', linestyle='--', linewidth=2, 
           label=f'μ + 3σ = {upper_threshold:.2f}')

# 添加图例和标签
plt.title('标准差法(3σ原则)异常值检测')
plt.xlabel('数据值')
plt.ylabel('频数')
plt.legend()
plt.grid(axis='y', linestyle='--', alpha=0.7)

# 添加颜色说明
blue_patch = mpatches.Patch(color='skyblue', label='正常数据')
red_patch = mpatches.Patch(color='red', label='异常值区域')
plt.legend(handles=[blue_patch, red_patch, 
                   plt.Line2D([], [], color='green', linestyle='-', label=f'均值 (μ)'),
                   plt.Line2D([], [], color='orange', linestyle='--', label=f'3σ 边界')])

plt.show()
    